subprojects/core/src/main/java/org/gradle/initialization/layout/BuildLayout.java

```diff
@@ -17,12 +17,15 @@
 
 import org.gradle.initialization.SettingsLocation;
 import org.gradle.internal.scripts.ScriptFileResolver;
+import org.gradle.internal.service.scopes.Scope;
+import org.gradle.internal.service.scopes.ServiceScope;
 
 import javax.annotation.Nullable;
 import java.io.File;
 
 import static org.gradle.initialization.DefaultProjectDescriptor.BUILD_SCRIPT_BASENAME;
 
+@ServiceScope({Scope.BuildSession.class, Scope.Build.class})
```

Comment:
🤔 Having it in both places looks fishy. I suppose, the settings file may come and go between builds in the same session? Then some splitting is due. Not saying it should be done in this PR.

Reply:
It's fishy indeed!

The build session one is exposed here:
https://github.com/gradle/gradle/blob/794be555f205535aba1d33b9f6bc56990ac92aab/subprojects/core/src/main/java/org/gradle/internal/service/scopes/CoreBuildSessionServices.java#L103-L106

I'd guess it's the one for the root of the build tree that can be determined early on, and maybe some other services already need to act on the "base" layout information.

The per-build scope is very clear, because each build in a build tree can have it's own location and settings file, etc.